# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

In [4]:
#Identify the dimensions of the dataset by determining the number of rows and columns it contains.
data_shape = df.shape
print(f"Shape of the DataFrame: {data_shape}")

Shape of the DataFrame: (4008, 11)


In [5]:
#Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 
#You should also provide suggestions for fixing any incorrect data types.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [6]:
df.nunique() #valores dentro de la variable

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [7]:
df["GENDER"].unique() #valores unicos dentro de la variable

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [8]:
df = df.replace(to_replace = ["Femal", "Male"], value = [["F"], "M"])

In [9]:
df = df.replace(to_replace=["female"], value="F")

In [10]:
df["GENDER"]= df["GENDER"].fillna(value = "ND")

In [11]:
df['GENDER'] = df['GENDER'].astype("string")

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     4008 non-null   string 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(7), string(1)
memory usage: 344.6+ KB


- Identify the number of unique values for each column and determine which columns appear to be categorical.
- You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

**VARIABLE QUALITATIVE**
- Customer: V categorical nominal
- ST: V categorical nominal
- GENDER: V categorical nominal
- Education: V categorical ordinal
- Policy Type: V categorical nominal
- Vehicle Class: V categorical nominal

In [13]:
df['Customer'].unique()
df['ST'].unique()
df['GENDER'].unique()
df['Education'].unique()
df['Policy Type'].unique()
df['Vehicle Class'].unique()

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury SUV', 'Sports Car',
       'Luxury Car', nan], dtype=object)

In [14]:
df['Customer'].unique()

array(['RB50392', 'QZ44356', 'AI49188', ..., 'CW49887', 'MY31220', nan],
      dtype=object)

In [15]:
df['ST'].unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [16]:
df['GENDER'].unique()

<StringArray>
['ND', 'F', 'M']
Length: 3, dtype: string

In [17]:
df['Education'].unique()

array(['Master', 'Bachelor', 'High School or Below', 'College',
       'Bachelors', 'Doctor', nan], dtype=object)

In [18]:
df['Policy Type'].unique()

array(['Personal Auto', 'Corporate Auto', 'Special Auto', nan],
      dtype=object)

In [19]:
df['Vehicle Class'].unique()

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury SUV', 'Sports Car',
       'Luxury Car', nan], dtype=object)

In [21]:
#Conclusion categorical:
categorical_variables = df.describe(include='object')
categorical_variables

Customer      ST Education Customer Lifetime Value  \
count      1071    1071      1071                    1068   
unique     1071       8         6                    1027   
top     RB50392  Oregon  Bachelor              445811.34%   
freq          1     320       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

**Conclusion categorical:**
- `Customer:` There are 1071 unique customers in the dataset. Each customer appears only once.
- `ST (State)`: There are 8 unique states in the data. Oregon seems to be the most common state, appearing 320 times in the dataset.
- `Education:` There are 6 unique education levels in the data. The majority of customers seem to have a Bachelor's degree, with 324 customers reporting this education level.
- `Policy Type:` There are 3 different policy types in the data. The majority of customers have a Personal Auto policy, with 576 customers reporting this policy type.
- `Vehicle Class:` There are 6 different vehicle classes in the data. The most common vehicle class appears to be the Four-Door Car, with 780 customers reporting this vehicle type.

**VARIABLE QUANTITATIVE**
- Customer Lifetime Value: 
- Income: V quantitative 
- Monthly Premium Auto: 
- Number of Open Complaints:
- Total Claim Amount:

In [22]:
#Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. 
df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [23]:
df['Income'].mode()

0    0.0
Name: Income, dtype: float64

In [24]:
df['Monthly Premium Auto'].mode()

0    65.0
Name: Monthly Premium Auto, dtype: float64

In [25]:
df['Total Claim Amount'].mode()

0    321.6
Name: Total Claim Amount, dtype: float64

**Conclusion quantitative:**
`Income:`
- Count: There are 1071 entries for income.
- Mean: The average income among the customers is approximately 39,295.70.
- Standard Deviation: The income values have a relatively high standard deviation of approximately 30,469.43, indicating a wide dispersion from the mean.
- Minimum: The lowest reported income is 0.00.
- 25th Percentile: 25% of customers have an income of 14,072.00 or lower.
- Median (50th Percentile): The median income, or the middle value when all incomes are sorted, is 36,234.00.
- 75th Percentile: 75% of customers have an income of 64,631.00 or lower.
- Maximum: The highest reported income is 99,960.00.
- Mode: The mode, is $0.0. This means that the value of 0.0 occurs most frequently

`Monthly Premium Auto:`
- Count: There are 1071 entries for monthly premium auto.
- Mean: The average monthly premium auto among the customers is approximately 193.23.
- Standard Deviation: The values for monthly premium auto have a relatively high standard deviation of approximately 1601.19, indicating a wide dispersion from the mean.
- Minimum: The lowest reported monthly premium auto is 61.00.
- 25th Percentile: 25% of customers have a monthly premium auto of 68.00 or lower.
- Median (50th Percentile): The median monthly premium auto, or the middle value when all premiums are sorted, is 83.00.
- 75th Percentile: 75% of customers have a monthly premium auto of 109.50 or lower.
- Maximum: The highest reported monthly premium auto is 35,354.00.
- Mode: The mode is 65.0. This means that the value of 65.0 occurs most frequently

`Total Claim Amount`
- Count: There are 1071 entries for the total claim amount.
- Mean: The average total claim amount among the customers is approximately 404.99.
- Standard Deviation: The values for the total claim amount have a standard deviation of approximately 293.03, indicating a spread of claim amounts around the mean.
- Minimum: The lowest reported total claim amount is 0.38.
- 25th Percentile: 25% of customers have a total claim amount of 202.16 or lower.
- Median (50th Percentile): The median total claim amount, or the middle value when all claims are sorted, is 354.73.
- 75th Percentile: 75% of customers have a total claim amount of 532.80 or lower.
- Maximum: The highest reported total claim amount is 2893.24.
- Mode: The mode is $321.6. This means that the value of 321.6 occurs most frequently

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [27]:
#Top 5 Less Common Customer Locations
customer_orders = df['ST'].value_counts().nsmallest(5)
less_common_locations = pd.Series(customer_orders)

print(less_common_locations)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [28]:
# Your code here
type_of_policy = df['Policy Type'].value_counts()
policy_type = pd.Series(type_of_policy)
policy_type

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [29]:
max_policy = type_of_policy.idxmax() #maximo de una serie
max_policy

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [30]:
# Create a DataFrame containing only Personal Auto policies
personal_auto_df = df.loc[df['Policy Type'] == 'Personal Auto']
personal_auto_df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington     ND                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
4  GA49547  Washington      M  High School or Below              536307.65%   
5  OC83172      Oregon      F              Bachelor              825629.78%   

    Income  Monthly Premium Auto Number of Open Complaints    Policy Type  \
0      0.0                1000.0                    1/0/00  Personal Auto   
1      0.0                  94.0                    1/0/00  Personal Auto   
2  48767.0                 108.0                    1/0/00  Personal Auto   
4  36357.0                  68.0                    1/0/00  Personal Auto   
5  62902.0                  69.0                    1/0/00  Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
4  Four-Door Car           17.269323  
5   Two-Door Car          159.383042

In [37]:
# Calculate the average income for customers with Personal Auto policies
average_income_personal_auto = personal_auto_df['Income'].mean()
average_income_personal_auto

38180.69871794872

In [38]:
# Create a DataFrame containing only Corporate Auto policies
corporate_auto_df = df.loc[df['Policy Type'] == 'Corporate Auto']
corporate_auto_df.head()

Customer          ST GENDER Education Customer Lifetime Value   Income  \
3   WW63253  California      M  Bachelor              764586.18%      0.0   
6   XZ87318      Oregon      F   College              538089.86%  55350.0   
7   CF85061     Arizona      M    Master              721610.03%      0.0   
8   DY87989      Oregon      M  Bachelor             2412750.40%  14072.0   
12  DP39365  California     ND    Master              879879.70%  77026.0   

    Monthly Premium Auto Number of Open Complaints     Policy Type  \
3                  106.0                    1/0/00  Corporate Auto   
6                   67.0                    1/0/00  Corporate Auto   
7                  101.0                    1/0/00  Corporate Auto   
8                   71.0                    1/0/00  Corporate Auto   
12                 110.0                    1/2/00  Corporate Auto   

    Vehicle Class  Total Claim Amount  
3             SUV          529.881344  
6   Four-Door Car          321.600000  
7   Four-Door Car          363.029680  
8   Four-Door Car          511.200000  
12  Four-Door Car          472.029737

In [39]:
# Calculate the average income for customers with Corporate Auto policies
average_income_corporate_auto = corporate_auto_df['Income'].mean()
average_income_corporate_auto

41390.31196581197

In [40]:
print("Average Income for Personal Auto:", average_income_personal_auto)
print("Average Income for Corporate Auto:", average_income_corporate_auto)

Average Income for Personal Auto: 38180.69871794872
Average Income for Corporate Auto: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here